Add font thumbnail preview support

Thesacraft opened the pull request:
I added simple font thumbnail previews for .ttf, .otf and .woff files like it was requested in issue #304 . Currently there are three text sizes (10,12,15) that i selected. I stacked the sample text in the different font sizes next to each other and tried to dynamically fit them next to each other. I am open to changing how the text alignment works and other changes. At the moment it looks like this:
![font support](https://github.com/TagStudioDev/TagStudio/assets/91694323/36470bb3-e919-47a8-84e1-02680c29112a)
![font_support_2](https://github.com/TagStudioDev/TagStudio/assets/91694323/9eebbfe7-5e62-41bc-8553-d7cad29f35e4)

Thesacraft commented:
I added a bit of information about the font as requested:
![grafik](https://github.com/TagStudioDev/TagStudio/assets/91694323/906d5701-5361-4c8c-9292-be587ee0bff4)
I could not find how to get the weight of the font, if someone knows how to do that let me know.

olipra commented:
This is a very good start! Already miles better than the file icon thumbnail. I'm just wondering whether it's worth aligning the text vertically instead of horizontally, and I'm also wondering if displaying the font in different sizes actually is worth it considering the small space it has to fit it. It's less of a problem for the bigger preview but it makes fonts in thumbnails hard to recognize, and I guess the thumbnail and preview can't show two different things. Otherwise it would make more sense for the thumbnail to only show "Aa" or something like that.

Thesacraft commented:
I changed it in the way you described here is the result:
![grafik](https://github.com/TagStudioDev/TagStudio/assets/91694323/d4d62452-609f-4757-a239-bc1e8a57727a)
I also adjusted the font sizes to 12,15,20

olipra commented:
This looks perfect to me! This is gonna be a life-changing feature for me and hopefully other type enthusiasts haha 🙌 Also thank you for taking my feature request in consideration and working on it!

CyanVoxel commented on tagstudio/src/qt/helpers/text_wrapper.py:
```diff
@@ -0,0 +1,54 @@
+    if draw is None:
+        bg = Image.new("RGB", (width, width), color="#1e1e1e")
+        draw = ImageDraw.Draw(bg)
+    if draw.textlength(text, font=font) > 256:
+        # print(draw.textlength(text, font=font))
+        for i in range(
+            int(len(text) / int(draw.textlength(text, font=font)) * 256) - 2,
+            0,
+            -1,
+        ):
+            if draw.textlength(text[:i], font=font) < 256:
+                # print(len(text))
+                # print(i)
```
Change to use `width` variable instead of hardcoded values (plus remove print comments)
```suggestion
    if draw.textlength(text, font=font) > width:
        for i in range(
            int(len(text) / int(draw.textlength(text, font=font)) * width) - 2,
            0,
            -1,
        ):
            if draw.textlength(text[:i], font=font) < width:
                return i
```

CyanVoxel commented on tagstudio/src/qt/helpers/text_wrapper.py:
```diff
@@ -0,0 +1,54 @@
+    i = 0
+    last_i = 0
+    while wrap_line(text[i:], font=font, width=width, draw=draw) > 0:
+        i = wrap_line(text[i:], font=font, draw=draw) + last_i
```
Pass width variable to `wrap_line`
```suggestion
        i = wrap_line(text[i:], font=font, width=width, draw=draw) + last_i
```

CyanVoxel commented on tagstudio/src/core/constants.py:
```diff
@@ -7,6 +7,11 @@
+FONT_SAMPLE_TEXT: str = (
+    """ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789!?@$%(){}[]"""
+)
+FONT_SAMPLE_SIZES: list[int] = [12, 15, 20]
```
Slight adjustment to font sample sizes to show more difference between them
```suggestion
FONT_SAMPLE_SIZES: list[int] = [10, 15, 20]
```

CyanVoxel commented on tagstudio/src/core/constants.py:
```diff
@@ -107,6 +112,7 @@
 ]
 PROGRAM_TYPES: list[str] = [".exe", ".app"]
 SHORTCUT_TYPES: list[str] = [".lnk", ".desktop", ".url"]
+FONT_TYPES: list[str] = [".ttf", ".otf", ".woff"]
```
Add `.ttc` and `.woff2` file extensions to the constants list (I haven't personally verified `.woff2` but `.ttc` seems to work fine)
```suggestion
FONT_TYPES: list[str] = [".ttf", ".ttc", ".otf", ".woff", ".woff2"]
```

CyanVoxel commented on tagstudio/src/qt/widgets/thumb_renderer.py:
```diff
@@ -183,7 +187,33 @@ def render(
+                    image = bg
+                # Fonts ========================================================
+                elif _filepath.suffix.lower() in FONT_TYPES:
+                    if gradient:
+                        # handles small thumbnails
+                        bg = Image.new("RGB", (256, 256), color="#1e1e1e")
+                        draw = ImageDraw.Draw(bg)
+                        font = ImageFont.truetype(_filepath, size=170)
+                        draw.text((10, 0), "Aa", font=font)
+                    else:
+                        # handles big thumbnails and renders a sample text in multiple font sizes
+                        bg = Image.new("RGB", (256, 256), color="#1e1e1e")
+                        draw = ImageDraw.Draw(bg)
+                        lines_of_padding = 2
+                        y_offset = 0
+
+                        for font_size in FONT_SAMPLE_SIZES:
+                            font = ImageFont.truetype(_filepath, size=font_size)
+                            text_wrapped: str = wrap_full_text(
+                                FONT_SAMPLE_TEXT, font=font, draw=draw
```
Use native preview sizes with dynamic scaling
```suggestion
                    # Scale the sample font sizes to the preview image
                    # resolution,assuming the sizes are tuned for 256px.
                    scaled_sizes: list[int] = [
                        math.floor(x * (adj_size / 256)) for x in FONT_SAMPLE_SIZES
                    ]
                    if gradient:
                        # handles small thumbnails
                        bg = Image.new("RGB", (adj_size, adj_size), color="#1e1e1e")
                        draw = ImageDraw.Draw(bg)
                        font = ImageFont.truetype(
                            _filepath, size=math.ceil(adj_size * 0.65)
                        )
                        draw.text((10, 0), "Aa", font=font)
                    else:
                        # handles big thumbnails and renders a sample text in multiple font sizes
                        bg = Image.new("RGB", (adj_size, adj_size), color="#1e1e1e")
                        draw = ImageDraw.Draw(bg)
                        lines_of_padding = 2
                        y_offset = 0

                        for font_size in scaled_sizes:
                            font = ImageFont.truetype(_filepath, size=font_size)
                            text_wrapped: str = wrap_full_text(
                                FONT_SAMPLE_TEXT, font=font, width=adj_size, draw=draw
```

CyanVoxel reviewed (changes requested):
Amazing work on this! I'm really happy with how this has come out. I've just got a few suggestions that come down to adding dynamic scaling with different preview resolutions and a couple extra file types.

Thank you so much for working on this!
